### PR TITLE
Add support for captive portals

### PR DESCRIPTION
**Pull request by guyonvarch**
The Python browser view was only able to cycle between play and
settings with a shortcut.

This commit adds the following:

- periodically check the connection, and update:
  - connection status,
  - and captive_portal_url: String.
- show a “Go to captive portal button” in case the check detect a captive
portal, and update the url of the settings to show it in case the button
is clicked.

## TODO

- [x] Tested on VM
- [x] Tested on a real machine

### CODE REVIEW

**Comment by knuton** [comment]
Good call to pull the captive portal check into the kiosk browser.

The URLs the kiosk cycles through were kept abstract on purpose and I am not sure this needs to change; the kiosk does not know about Play or the PlayOS controller. The kiosk browser might just check for captive portals no matter which URL it is at, and always show the prompt to open the portal if it detects one. This is also how browsers behave.

In fact, we could consider the kiosk browser just a weird tabbed browser where you cycle through the tabs using Ctrl-Shift-F12.

File: kiosk/kiosk_browser/connection.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+            try:
+                r = requests.get(check_connection_url, allow_redirects = False)
+
+                if r.status_code == 200:
+                    self._is_connected = True
+                    self._set_captive_portal_url('')
+                elif r.status_code in [301, 302, 303, 307]:
+                    self._is_connected = False
+                    self._set_captive_portal_url(r.headers['Location'])
+                else:
```
Comment (knuton) [inline kiosk/kiosk_browser/connection.py]: I wonder if an additional case for `511 Network Authentication Required` is needed: https://tools.ietf.org/html/rfc6585#section-6

In the case, I guess the captive portal URL would need to be the test URL itself.

Comment (knuton) [reply kiosk/kiosk_browser/connection.py]: I could not find good data on whether this is used much in practice. Firefox does not seem to support it, based on this issue: https://bugzilla.mozilla.org/show_bug.cgi?id=728658

**Comment by knuton** [comment]
I looked into covering more forms of captive portals.

My first impulse is, it's not worth it. Part of what OS and browser vendors have to deal with, is the crookery of malicious network providers:

- https://www.tanaza.com/blog/ios-7-and-captive-portal-a-guide-to-captive-portal-requirements/
- https://news.ycombinator.com/item?id=13372428
- https://github.com/httpwg/wiki/wiki/Captive-Portals

It should be safe to assume, that only overall reasonable networks should be used with Play Computers. Users should not be expected to use airport or hotel wifi or worse for a PlayOS installation.

In particular, we don't want to deal with:

- ad-banner portals,
- short-lived wifi sessions,
- shady networks trying to fool the device into thinking it is online (to avoid the captive portal being loaded in a sandboxed browser),
- captive portals delivering the portal at the test URL (instead of delivering a redirect).

The technique employed here should cover most legitimate cases, such that if it doesn't work at an installation site it should be fair for support to say "The wifi network you are trying to use is not adequate for a persistent kiosk installation, please look into sourcing a proper network to use our device".

Basically what we enable here is a short-cut to use PlayOS in wifi networks that work by white-listing MAC addresses for extended periods of time (and could probably be configured by IT staff instead, though they might be hard to get hold off).

File: kiosk/kiosk_browser/connection.py (hunk 1 above)
Comment (guyonvarch) [reply kiosk/kiosk_browser/connection.py]: Do you know if we could get information concerning captive portals actually used ? If every captive portal already used gives a redirect code and a location, or if some of them gives back a 511.

File: kiosk/kiosk_browser/connection.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        thread = threading.Thread(target=self._check, args=[])
+        thread.daemon = True
+        thread.start()
+
+    def is_connected(self):
+        return self._is_connected
+
+    def _check(self):
+        # Initial sleep to be sure _set_captive_portal_url is accessible
+        time.sleep(1)
```
[the remark this answers is not in the document]
Comment (guyonvarch) [reply kiosk/kiosk_browser/connection.py]: I moved the connection daemon start at the end of the constructor, it may be enough. I’ll test this on a virtual machine and on a real machine.

File: kiosk/kiosk_browser/connection.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                    self._is_connected = True
+                    self._set_captive_portal_url('')
+                elif r.status_code in [301, 302, 303, 307]:
+                    self._is_connected = False
+                    self._set_captive_portal_url(r.headers['Location'])
+                else:
+                    self._is_connected = False
+                    self._set_captive_portal_url('')
+
+            except requests.ConnectionError as e:
```
[the remark this answers is not in the document]
Comment (guyonvarch) [reply kiosk/kiosk_browser/connection.py]: Good catch! I successfully simulated a timeout exception, it was effectively not caught. It made only the connection daemon to crash though, not the entire kiosk browser.

File: kiosk/kiosk_browser/connection.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import requests
+import tempfile
+import threading
+import time
+
+check_connection_url = 'http://captive.apple.com/'
+sleep_time = 5
```
[the remark this answers is not in the document]
Comment (guyonvarch) [reply kiosk/kiosk_browser/connection.py]: Good idea, let’s start with 60 seconds if the connection is down.

File: kiosk/kiosk_browser/connection.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (guyonvarch) [reply kiosk/kiosk_browser/connection.py]: Now that this branch does not trigger the captive portal, It’s ok to reset the url.

We could also show a message to the user in this case, something like “There is some trouble with the connection right now”. But it could be a false positive in case there is something unpredictable going on the captive url. Or, the application itself could already show an error message if the connection is really down.

Now, in case the user is behind a captive portal, and this captive portal use a method not handled here, it could be interesting.

With pros and cons, I would keep it as it is.

[the remark this answers is not in the document]
Comment (knuton) [reply kiosk/kiosk_browser/connection.py]: I agree, let's go with this implementation and observe.

File: kiosk/kiosk_browser/captive_portal_message.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+class CaptivePortalMessage(QWidget):
+
+    def __init__(self, press_button):
+        super(CaptivePortalMessage, self).__init__()
+
+        self._label = QLabel()
+
+        self._button = QPushButton()
+        self._button.setFixedWidth(180)
```
[the remark this answers is not in the document]
Comment (guyonvarch) [reply kiosk/kiosk_browser/captive_portal_message.py]: Well spotted, I removed fixed sizes.

**Comment by guyonvarch** [comment]
> Also a remaining question: Is it responsible to use `captive.apple.com` here?
> 
> The main arguments for using it are
> 
> it is convenient,
> a lot is riding on it, so it is likely to be supported for a long time.
> The main arguments against using it are
> 
> privacy concerns,
> giving away control.

I would opt against using `captive.apple.com`, I think privacy concerns is more important than convenience.

We would just have to serve a page, say `captive.dividat.com`, returning a 200 status.

**Comment by knuton** [comment]
> We would just have to serve a page, say `captive.dividat.com`, returning a 200 status.

"Just", and it's an added potential source of failure/confusion for live installations in case of any operations mishaps or infrastructure failure. :slightly_smiling_face: 

Still, this is the right thing to do. Can you change the hostname?

**Comment by guyonvarch** [comment]
I tested back the captive portal connection on a real machine with `captive.apple.com`, it worked as expected.

**Comment by knuton** [comment]
Very well, I think this is about ready.

Have you convinced yourself of the impossibility of race conditions on the URL callback (https://github.com/dividat/playos/pull/16#discussion_r320821321)? Defensive programming is also always an option.

**Comment by guyonvarch** [comment]
The bug is fixed now, but we’d better use defensive programming, in case there are changes afterward. I protected the main loop of the connection daemon.
